op-node/rollup/derive/span_channel_out.go

```diff
@@ -77,115 +114,119 @@ func (co *SpanChannelOut) AddBlock(rollupCfg *rollup.Config, block *types.Block)
+		return uint64(co.lastCompressedRLPSize), nil
 	}
-	if co.compress.FullErr() != nil {
-		err = co.compress.FullErr()
+
+	// we must compress the data to check if we've met or exceeded the target size
+	if err = co.compress(); err != nil {
+		return 0, err
+	}
+	co.lastCompressedRLPSize = co.activeRLP().Len()
+
+	// if the channel is now full, either return the compressed data, or the compressed previous data
+	if err := co.FullErr(); err != nil {
+		// if there is only one batch in the channel, it *must* be returned
 		if len(co.spanBatch.Batches) == 1 {
-			// Do not return ErrCompressorFull for the first block in the batch
-			// In this case, reader must be empty. then the contents of compressor will be copied to reader when the channel is closed.
-			err = nil
+			return uint64(co.compressed.Len()), nil
+		}
+
+		// if there is more than one batch in the channel, we revert the last batch
+		// by switching the RLP buffer and doing a fresh compression
+		co.swapRLP()
+		if err = co.compress(); err != nil {
+			return 0, err
 		}
-		// If there are more than one blocks in the channel, reader should have data that preserves previous compression result before adding this block.
-		// So, as a result, this block is not added to the channel and the channel will be closed.
-		return uint64(written), err
+		// co.FullErr should always return FullErr by this point
+		return uint64(co.compressed.Len()), co.FullErr()
 	}
 
-	// If compressor is not full yet, reader must be reset to avoid submitting invalid frames
-	co.reader.Reset()
-	return uint64(written), err
+	return uint64(co.compressed.Len()), nil
+}
+
+// compress compresses the active RLP buffer and checks if the compressed data is over the target size.
+// it resets all the compression buffers because Span Batches aren't meant to be compressed incrementally.
+func (co *SpanChannelOut) compress() error {
+	co.compressed.Reset()
+	co.compressor.Reset(co.compressed)
+	if _, err := co.compressor.Write(co.activeRLP().Bytes()); err != nil {
+		return err
+	}
+	co.compressor.Flush()
+	co.checkFull()
+	return nil
 }
 
 // InputBytes returns the total amount of RLP-encoded input bytes.
 func (co *SpanChannelOut) InputBytes() int {
-	return co.rlpLength
+	return co.activeRLP().Len()
 }
 
-// ReadyBytes returns the number of bytes that the channel out can immediately output into a frame.
-// Use `Flush` or `Close` to move data from the compression buffer into the ready buffer if more bytes
-// are needed. Add blocks may add to the ready buffer, but it is not guaranteed due to the compression stage.
+// ReadyBytes returns the total amount of compressed bytes that are ready to be output.
+// Span Channel Out does not provide early output, so this will always be 0 until the channel is closed or full
 func (co *SpanChannelOut) ReadyBytes() int {
-	return co.reader.Len()
+	if co.closed || co.FullErr() != nil {
+		return co.compressed.Len()
+	}
+	return 0
```

Comment:
Interesting, almost looks like an oversight in the old implementation?

Reply:
The old implementation did a lot of extra work to reset the read buffer, which I think was to ensure the reader was empty when this would get called. See comments in the old AddSingularBatch `// Must reset reader to avoid partial output`